fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/directory/LuceneSerializer.java

```diff
@@ -0,0 +1,189 @@
+import org.apache.lucene.codecs.compressing.CompressionMode;
+import org.apache.lucene.codecs.compressing.Compressor;
+import org.apache.lucene.codecs.compressing.Decompressor;
+import org.apache.lucene.store.ByteArrayDataInput;
+import org.apache.lucene.store.ByteBuffersDataOutput;
+import org.apache.lucene.util.BytesRef;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import java.util.Arrays;
+
+public class LuceneSerializer {
+    private static final Logger LOGGER = LoggerFactory.getLogger(LuceneSerializer.class);
+    private static final int ENCODING_ENCRYPTED = 1;
+    private static final int ENCODING_COMPRESSED = 2;
+    private static final byte COMPRESSION_VERSION_FOR_HIGH_COMPRESSION = 0;
+
+    private static class EncodingState {
+        private boolean compressed;
+        private boolean encrypted;
+
+        private EncodingState() {
+            this.compressed = false;
+            this.encrypted = false;
+        }
+
+        private void setCompressed(boolean compressed) {
+            this.compressed = compressed;
+        }
+
+        private void setEncrypted(boolean encrypted) {
+            this.encrypted = encrypted;
+        }
+
+        private boolean isCompressed() {
+            return compressed;
+        }
+
+        private boolean isEncrypted() {
+            return encrypted;
+        }
+    }
+
+    @Nullable
+    public static byte[] encode(@Nonnull String resourceDescription, @Nullable byte[] data, boolean compress, boolean encrypt) {
+        if (data == null) {
+            return null;
+        }
+        final EncodingState state = new EncodingState();
+        if (compress) {
+            data = compress(data, state);
+        }
+        if (encrypt) {
+            data = encrypt(data, state);
+        }
+        final byte[] encoded = new byte[data.length + 1];
+        int code = 0;
+        if (state.isCompressed()) {
+            code = code | ENCODING_COMPRESSED;
+        }
+        if (state.isEncrypted()) {
+            code = code | ENCODING_ENCRYPTED;
+        }
+        encoded[0] = (byte) code;
+        System.arraycopy(data, 0, encoded, 1, data.length);
+        LOGGER.trace("Encoded data for resourceDescription={}, compressionSupposed={}, encryptionSupposed={}, compressedEventually={}, encryptedEventually={}, originalDataSize={}, encodedDataSize={}",
+                resourceDescription, compress, encrypt, state.isCompressed(), state.isEncrypted(), data.length, encoded.length);
+        return encoded;
+    }
+
+    @Nullable
+    public static byte[] decode(@Nonnull String resourceDescription, @Nullable byte[] data) {
+        if (data == null) {
+            return null;
+        }
+        if (data.length < 2) {
+            throw new RecordCoreException("Invalid data")
+                    .addLogInfo(LogMessageKeys.DATA_VALUE, data);
+        }
+
+        final byte encoding = data[0];
```

Comment:
Usually when you are compressing or encrypting you act on streams of data vs. byte[].  If you do that, you lose the extra array copy.